### DISCUSSION TITLE
Migrate existing etcd to mysql

### DISCUSSION THREAD

**Discussion by febri4n** [k3s-io/k3s#9530] [Q&A] [answered]
**First question**
I have 6 ha k3s nodes, 3 controlplane with embedded etcd and 3 workers. is it possible to migrate from embedded etcd to external db mysql?
Here are the nodes that I have.
NAME           STATUS   ROLES                                          AGE     VERSION
k3s-master-1   Ready    control-plane,etcd,master   3d23h   v1.24.17+k3s1
k3s-master-2   Ready    control-plane,etcd,master   3d23h   v1.24.17+k3s1
k3s-master-3   Ready    control-plane,etcd,master   3d23h   v1.24.17+k3s1
k3s-worker-1   Ready    <none>                                    3d23h   v1.24.17+k3s1
k3s-worker-2   Ready    <none>                                    3d23h   v1.24.17+k3s1
k3s-worker-3   Ready    <none>                                    3d23h   v1.24.17+k3s1

**Second question**
Does adding 1 more node with datastore-endpoint to mysql include moving the three previous etcd to the new node?
or I just need to add datastore endpoint --datastore-endpoint="mysql://username:password@tcp(hostname:3306)/database-name" to /etc/systemd/system/k3s.service (existing node)?

**Comment by brandond** [comment, marked as answer]
> is it possible to migrate from embedded etcd to external db mysql?

Not directly, no. You would have to manually migrate your data from etcd to kine using 3rd party etcd clients. The only direct migration supported by k3s is from embedded sqlite to embedded etcd, by adding --cluster-init to an existing server node.

**Reply by febri4n** [reply]
Thank you for your response, but i want to ask again. 
This my ha k3s
![image](https://github.com/k3s-io/k3s/assets/18482250/b668feb7-5afd-463c-bffc-661cdd68ef15)

i create my new instance, install mysql on it. and then i execute this command:

curl -sfL https://get.k3s.io | INSTALL_K3S_VERSION=v1.24.17+k3s1 K3S_TOKEN=SECRET sh -s - server \
--datastore-endpoint="mysql://mysqluser:mysqlpassword@tcp(mysqlhost:3306)/k3s_database" \
--server https://k3s-master-1-ip:6443


this is the result
![image](https://github.com/k3s-io/k3s/assets/18482250/3ede3491-800c-4bc2-89ad-a69a272a0902)
node **ha-k3s-mysql** only showing when i execute **kubectl get node** only in **node ha-k3s-mysql** 


Please help to confirm, is this correct, that I am now using an external db with mysql?
when i exec **kubectl get node** in **k3s-master-1**, node **ha-k3s-mysql** not showing. 

![image](https://github.com/k3s-io/k3s/assets/18482250/e765cec2-a987-4682-95a6-83ba791a99ff)

**Reply by brandond** [reply]
> i create my new instance, install mysql on it. and then i execute this command:

Your use of the mysql datastore endpoint overrides your request to join the cluster. You cannot BOTH use an external datastore, AND join an etcd cluster.

> when i exec kubectl get node in k3s-master-1, node ha-k3s-mysql not showing.

You have two different clusters. One of them is using etcd. one of them is using mysql. You cant have a single cluster that uses two different datastores simultaneously.
